Src/FluentAssertions/Equivalency/GenericDictionaryEquivalencyStep.cs

```diff
@@ -154,15 +154,81 @@ private static Type GetIDictionaryInterface(Type expectedType)
             return GetIDictionaryInterfaces(expectedType).Single();
         }
 
-        private static bool AssertSameLength<TSubjectKey, TSubjectValue, TExpectKey, TExpectedValue>(
+        private static bool AssertSameLength<TSubjectKey, TSubjectValue, TExpectedKey, TExpectedValue>(
             IDictionary<TSubjectKey, TSubjectValue> subject,
-            IDictionary<TExpectKey, TExpectedValue> expectation)
+            IDictionary<TExpectedKey, TExpectedValue> expectation) where TExpectedKey : TSubjectKey
```

Comment:
Doesn't this introduce "asymmetry"?
E.g. if `TSubjectKey` derives from `TExpectedKey` instead, then `AssertSameLength(dict1, dict2)` would work, but not `AssertSameLength(dict2, dict1)`.


How about when `TSubjectKey` and `TExpectedKey` are completely different types?
Is this a too strict constraint?

I hope my thoughts make any sense.

Reply:
Yes, it does introduce asymmetry but it is the same asymmetry that is already enforced in AssertIsCompatiblyTypedDictionary which is executed just before AssertSameLength. I added two tests that document the asymmetric behaviour. Additionally, I explained the asymmetry issue just below the `where TExpectedKey : TSubjectKey`

Thanks for your input. Please, let me know if my explanation makes sense.